src/pointcloud.rs

```diff
@@ -0,0 +1,154 @@
+use crate::config::PointCloud2ListenerConfig;
+use byteorder::{ByteOrder, LittleEndian};
+use colorgrad;
+use std::sync::{Arc, RwLock};
+
+use nalgebra::geometry::Point3;
+use tui::style::Color;
+
+use crate::transformation::ros_transform_to_isometry;
+use rosrust;
+use rustros_tf;
+
+pub struct PointCloud2Listener {
+    pub config: PointCloud2ListenerConfig,
+    pub points: Arc<RwLock<Vec<ColoredPoint>>>,
+    _tf_listener: Arc<rustros_tf::TfListener>,
+    _static_frame: String,
+    _subscriber: rosrust::Subscriber,
+}
+
+#[derive(Clone)]
+pub struct ColoredPoint {
+    pub point: Point3<f64>,
+    pub color: Color,
+}
+
+impl ColoredPoint {
+    pub fn new(point: Option<Point3<f64>>, color: Option<Color>) -> ColoredPoint {
+        ColoredPoint {
+            point: point.unwrap_or(Point3::new(0.0, 0.0, 0.0)),
+            color: color.unwrap_or(Color::Red),
+        }
+    }
+}
+
+pub fn get_channel_offset(name: &str, fields: &Vec<rosrust_msg::sensor_msgs::PointField>) -> u32 {
+    for field in fields {
+        if field.name == name {
+            return field.offset;
+        }
+    }
+    panic!("Could not find field {:}", name);
+}
+
+pub fn read_f32(bytes: &Vec<u8>, idx: u32) -> f32 {
+    LittleEndian::read_f32(&bytes[idx as usize..(idx + 4) as usize])
+}
+
+pub fn read_xyz(msg: &rosrust_msg::sensor_msgs::PointCloud2) -> Vec<Point3<f64>> {
+    let n_pts = msg.width * msg.height;
+    let mut points: Vec<Point3<f64>> = Vec::with_capacity(n_pts as usize);
+    let x_offset = get_channel_offset("x", &msg.fields);
+    let y_offset = get_channel_offset("y", &msg.fields);
+    let z_offset = get_channel_offset("z", &msg.fields);
+    for i in 0..n_pts {
+        let idx = i * msg.point_step;
+        points.push(Point3::new(
+            read_f32(&msg.data, idx + x_offset) as f64,
+            read_f32(&msg.data, idx + y_offset) as f64,
+            read_f32(&msg.data, idx + z_offset) as f64,
+        ));
+    }
+    points
+}
+
+pub fn colorize_from_rgb(
```

Comment:
there are 2 ways to store rgb in pointcloud2.  A single field with an f32 value (each color one byte) . Or as seperated fields.  I took the lazy way and implemented only what i encountered. This will panic and remind us if we need it